File: 卡码/代码随想录/Python/动态规划/02-背包问题/01-01背包/LeetCode_0494.py
#494. 目标和
from typing import *

class Solution:
  def findTargetSumWays(self, nums: List[int], target: int) -> int:
    total:int=sum(nums)
    if abs(target)>total or (total+target)%2==1:
      return 0
    n:int=(target+total)//2
    dp:List[int]=[0]*(n+1)
    dp[0]=1
    for i in range(len(nums)):
      for j in range(n,nums[i]-1,-1):
        dp[j]+=dp[j-nums[i]]
    return dp[n]

File: 卡码/代码随想录/Go/动态规划/02-背包问题/01-01背包/LeetCode_0494.go
// 494. 目标和
package main

import "math"

func findTargetSumWays(nums []int, target int) int {
	sum:=0
	for _,num:=range nums {
		sum+=num
	}
	if math.Abs(float64(target))>float64(sum)||(sum+target)%2==1 {
		return 0
	}
	n:=(sum+target)/2
	dp:=make([]int,n+1)
	dp[0]=1
	for i:=0;i<len(nums);i++ {
		for j:=n;j>=nums[i];j-- {
			dp[j]+=dp[j-nums[i]]
		}
	}
	return dp[n]
}
